### ISSUE TITLE
Keep stats about fixed errors

### ISSUE THREAD

**Issue by donaldpipowitch** [eslint/eslint#9147]
**Tell us about your environment**

* **ESLint Version:** 4.4.1
* **Node Version:** 6.9
* **npm Version:**

**What parser (default, Babel-ESLint, etc.) are you using?**

typescript-eslint-parser

**Please show your full configuration:**

[My config.](https://github.com/donaldpipowitch/typescript-experiments/blob/ce536d241ea9c065ddd41eb0955b39d640e572c1/eslint/.eslintrc.json)

**What did you do? Please include the actual source code causing the issue.**

I use the `CLIEngine` directly as my "Node API" to lint my code (see https://github.com/eslint/eslint/issues/9146 why). The code [looks like this](
https://github.com/donaldpipowitch/typescript-experiments/blob/ce536d241ea9c065ddd41eb0955b39d640e572c1/eslint/eslint.js).

When I add `options.fix = true;` before passing it to the `CLIEngine` my code gets fixed automatically when I call `executeOnFiles `. However it seems that I lost information about _what_ gets fixed. I'd like to keep that.

**What did you expect to happen?**

I'd expect something analogous to `fixableErrorCount` called `fixedErrorCount`. Ideally the `stats.results[].messages[]` would track fixed and non-fixed errors, too.

---

Thank you for the project! I hope my code snippet helps someone.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. There was a similar discussion about this in https://github.com/eslint/eslint/issues/7404 -- in short, while it might possible to track the number of autofixes applied when linting a given piece of code, that number might end up being confusing for the user. Due to the way ESLint applies fixes over multiple passes on the file, it might end up "fixing" issues that weren't present in the original file in the first place, and some "fixable" errors might not actually end up getting fixed on a given pass. This makes it nontrivial to track the number of fixed problems in a way that's actually useful.

Could you elaborate on what you would use the fixed error count for? That might help determine what method of counting fixed problems would be useful to you, and/or what alternatives might be available.

**Comment by donaldpipowitch** [comment]
You're right. https://github.com/eslint/eslint/issues/7404 sounds very similar. If you want to you can close this as a duplicate.

> Could you elaborate on what you would use the fixed error count for?

I basically just want to give feedback to the user that this command had some side effect on there source code and that _something_ has changed. (No detailed report. Just the number of total fixes.)

> Due to the way ESLint applies fixes over multiple passes ...

Yeah, I saw this in the source code. I wouldn't mind if the number is not 100% correct.  My feedback to the user is really more like _"Something has changed. We fixed 100 errors."_ IMHO it would be okay, if it fixed 90 or 110 instead...

**Comment by donaldpipowitch** [comment]
At the end of day I guess I basically want to know _if_ a file has changed or not. (It would be nice to know how many fixes were applied and which ones, but it is not that crucial for me.)

I just noticed I also need this information to actually write the fixed files back...?

I guess currently I'd need to something like this (`*Sync` file reading for this example):

```js
const { readFileSync, writeFileSync } = require('fs');
const { CLIEngine, linter } = require('eslint');

const patterns = ['./src/index.ts'];

const options = require('./.eslintrc.json');
options.fix = true;
const engine = new CLIEngine(options);

// interface Stats {
//   results: Array<{
//     filePath: string;
//     messages: Array<any>;
//     errorCount: number;
//     warningCount: number;
//     fixableErrorCount: number;
//     fixableWarningCount: number;
//     output: string;
//   }>;
//   errorCount: number;
//   warningCount: number;
//   fixableErrorCount: number;
//   fixableWarningCount: number;
// }
const stats = engine.executeOnFiles(patterns);

// fixes
let fixedFiles = 0;
stats.results.forEach(result => {
  const content = readFileSync(result.filePath, 'utf8');
  // this will be so slow for big files... :(
  if (content !== result.output) {
    writeFileSync(result.filePath, result.output);
    fixedFiles += 1;
  }
});
if (fixedFiles) {
  console.log(`Applied fixes to ${fixedFiles} file(s).`);
}

// errors
if (stats.errorCount) {
  console.error(`Found ${stats.errorCount} error(s).`);
  stats.results.forEach(result => {
    result.messages.forEach(msg =>
      console.log(`${result.filePath}:
    ${msg.line}:${msg.column} error ${msg.message}`)
    );
  });
}
```

**Comment by not-an-aardvark** [comment]
The `output` property is only present on the file results object if an autofix was applied for that file. So you can check for the existence of the `output` property to determine whether fixes were applied, without rereading the files from the filesystem.

**Comment by donaldpipowitch** [comment]
Good to know! Thank you.

**Comment by not-an-aardvark** [comment]
Should this issue be closed? It seems like checking for the existence of `output` solves your use case, but let me know if there's anything else I was missing.

**Comment by donaldpipowitch** [comment]
Yeah, I guess we can close it. Thanks.
